### ISSUE TITLE
NVC+OpenACC broken with SubDomainSet and MPI

### ISSUE THREAD

**Issue by FabioLuporini** [devitocodes/devito#1766]
Premise: This is a bug somewhere in the NVidia SDK (nvc+openacc), not a devito bug. But we need a workaround in devito. 
Explanation copy-pasted from slack follows.

Consider `SubDomainSet`'s `_bounds_glb_to_loc`. And consider the following (copy-pasted from a code comment):

> When the MPI-decomposed domains are very small and the thicknesses very large, it is possible that the combination of `x_m, x_M, x_ltkn, x_rtkn, ...` creates an iteration space with "negative size" -- for example, `x` starting at `x_m + x_ltkn = 3` and terminating at `x_M - x_rtkn = 0` (ie `size = -3` iterations). This simply means that ultimately no iterations will be executed -- any "negative size" would imply exactly this. 

Unfortunately, the NVC compiler/runtime for OpenACC has a bug when the iteration space size is smaller than `-1`. ...
Now, first of all: the bug has been confirmed and escalated to the NVidia compiler team (via @kenhester , thank you)
The issue crops up for us, as I was saying earlier, with `SubDomainSet` (or more generally with `MultiSubDomain`). In particular, it happens due to the following pair of lines: these, these, these, and these. These lines set the thicknesses to the loc_abs_max, so ultimately you can get a "negative size" smaller than -1, which triggers the bug (eventually computing wrong output).

Do we agree that if I set all `bounds_M[j]` to `1`, rather than to `dec.loc_abs_max`, we would obtain the same effect, ie no iterations executed as iteration space size systematically == -1, while also working around the NVC bug?
Why -- well we have the left thickness `bounds_m[j] = dec.loc_abs_max`, so for example we got x starting at `x_m + xi_ltkn = 0 + 3 = 3` and stopping at `x_M - xi_rtkn = 3 - 1 = 2`. The reason that here `x_M == 3` is because we're assuming that `dec.loc_abs_max = 3`

**Comment by FabioLuporini** [comment]
For completeness: the following MFE was extracted from a devito failing test and handed over to NVidia for reproducibility and to work out a patch:

```
#include <stdlib.h>
#include "openacc.h"
#include <stdio.h>

int main(int argc, char* argv[])
{
  int xi_ltkn = atoi(argv[1]);
  int xi_rtkn = atoi(argv[2]);
  int yi_ltkn = atoi(argv[3]);
  int yi_rtkn = atoi(argv[4]);
  int x_m = atoi(argv[5]);
  int x_M = atoi(argv[6]);
  int y_m = atoi(argv[7]);
  int y_M = atoi(argv[8]);

  printf("INPUT: %d %d %d %d  --  %d %d %d %d\n", xi_ltkn, xi_rtkn, yi_ltkn, yi_rtkn, x_m, x_M, y_m, y_M);
  printf("\n");

  acc_init(acc_device_nvidia);

  int* f_vec = (int*) malloc(sizeof(int)*4*4);
  int (*restrict f)[4] __attribute__ ((aligned (64))) = (int (*)[4]) f_vec;

  for (int i = 0; i < 4; i++) {
    for (int j = 0; j < 4; j++) {
      f[i][j] = 0;
    }
  }

  for (int i = 0; i < 4; i++) {
    for (int j = 0; j < 4; j++) {
      printf("%d ", f[i][j]);
    }
    printf("\n");
  }

  printf("\n");
  printf("======\n");
  printf("\n");

  #pragma acc enter data copyin(f[0:4][0:4])

  printf("xi from %d to %d\n", x_m + xi_ltkn, x_M - xi_rtkn);
  printf("yi from %d to %d\n", y_m + yi_ltkn, y_M - yi_rtkn);
  printf("Now offloading...\n");
  printf("\n");

  #pragma acc parallel loop collapse(2) present(f)
  for (int xi = x_m + xi_ltkn; xi <= x_M - xi_rtkn; xi += 1)
  {
    for (int yi = y_m + yi_ltkn; yi <= y_M - yi_rtkn; yi += 1)
    {
      f[xi][yi] = f[xi][yi] + 1;
    }
  }

  #pragma acc exit data copyout(f[0:4][0:4])
  #pragma acc exit data delete(f[0:4][0:4])

  for (int i = 0; i < 4; i++) {
    for (int j = 0; j < 4; j++) {
      printf("%d ", f[i][j]);
    }
    printf("\n");
  }
  printf("\n");

  return 0;
}
```

compile and run with:

```
 nvc -acc:gpu acc_bug.cpp -o acc_bug ; ./acc_bug 3 1 3 1 0 3 0 3  # ALL GOOD, it space size == -1
 nvc -acc:gpu acc_bug.cpp -o acc_bug ; ./acc_bug 3 2 3 2 0 3 0 3  # BROKEN, it space size < -1 (== -2 for each loop)
```

**Comment by FabioLuporini** [comment]
fixed by #1767
